gems/aws-crt-auth/lib/aws-crt-auth/credentials.rb

```diff
@@ -0,0 +1,103 @@
+# frozen_string_literal: true
+
+module Aws
+  module Crt
+    module Auth
+      # Utility class for Credentials.
+      # @api private
+      class Credentials
+        UINT64_MAX = 18_446_744_073_709_551_615
+
+        # @param [String] access_key_id
+        # @param [String] secret_access_key
+        # @param [String] session_token (nil)
+        # @param [Time|int] expiration (nil)
+        def initialize(access_key_id, secret_access_key,
+                       session_token = nil, expiration = nil)
+          if !access_key_id || access_key_id.empty?
+            raise ArgumentError, 'access_key_id  must be set'
+          end
+
+          if !secret_access_key || secret_access_key.empty?
+            raise ArgumentError, 'secret_access_key  must be set'
+          end
+
+          native = Aws::Crt.call do
+            Aws::Crt::Native.credentials_new(
+              access_key_id,
+              secret_access_key,
+              session_token,
+              expiration&.to_i || UINT64_MAX
+            )
+          end
+
+          @native = FFI::AutoPointer.new(native, self.class.method(:on_release))
+        end
+
+        # @return [FFI:Pointer]
+        attr_reader :native
+
+        # @return [String, nil]
+        def access_key_id
+          Aws::Crt::Native.credentials_get_access_key_id(@native) if @native
+        end
+
+        # @return [String, nil]
+        def secret_access_key
+          Aws::Crt::Native.credentials_get_secret_access_key(@native) if @native
+        end
+
+        # @return [String, nil]
+        def session_token
+          Aws::Crt::Native.credentials_get_session_token(@native) if @native
+        end
+
+        # @return [Time,nil]
+        def expiration
+          return unless @native
+
+          exp = Aws::Crt::Native.credentials_get_expiration(@native)
+          return if exp == UINT64_MAX
+
+          Time.at(exp)
+        end
+
+        # @return [Credentials]
+        def credentials
+          self
```

Comment:
what's the point of this function?

Reply:
Conforms to existing SDK credentials interface.  it is weird and pointless here.  But wherever we expect credentials, they can either be credentials or a credentials provider (which provides credentials with the `credentials` method).

Reply:
I don't think we should confuse the Ruby SDK's API with the CRT's API.
Credentials and CredentialsProviders are **not** interchangable in the CRT API. A CredentialsProvider can only fetch Credentials asynchronously. You won't be able to wrap that with a sync function like `credentials()` without risking deadlock

Reply:
Yeah - this makes sense.  I've removed the `credentials` and `set?` methods.